### PR TITLE
Add motion streak and particle 2d.

### PR DESCRIPTION
**Pull request by caryliu1999**
MotionStreak: https://github.com/cocos-creator/3d-tasks/issues/3757
ParticleSystem2D: https://github.com/cocos-creator/3d-tasks/issues/3271

Editor: https://github.com/cocos-creator/editor-3d/pull/4403

### CODE REVIEW

File: cocos/particle/index.ts [hunk 1]
```diff
@@ -9,13 +9,21 @@ import { ParticleSystem } from './particle-system';
 import { ParticleUtils } from './particle-utils';
 import CurveRange from './animator/curve-range';
 import { legacyCC } from '../core/global-exports';
+import { ParticleSystem2D } from './2d/particle-system-2d';
+import { MotionStreak } from './2d/motion-streak-2d';
+import { MotionStreakAssemblerManager } from './2d/motion-streak-2d-assembler';
+import { ParticleSystem2DAssembler } from './2d/particle-system-2d-assembler';
```
Comment (pandamicro) [inline cocos/particle/index.ts]: Please separate to `core/particle-2d` module

File: cocos/core/utils/misc.ts [hunk 2]
```diff
@@ -186,6 +187,52 @@ export function cloneable_DEV (obj) {
         ((obj.constructor && obj.constructor.prototype.hasOwnProperty('clone')) || obj.hasOwnProperty('clone'));
 }
 
+/**
+ * !#en Clamp a value between from and to.
+ * !#zh
```
Comment (pandamicro) [inline cocos/core/utils/misc.ts]: Please use correct document comment format

File: cocos/core/utils/misc.ts [hunk 3]
```diff
@@ -186,6 +187,52 @@ export function cloneable_DEV (obj) {
...
+    if (min_inclusive > max_inclusive) {
+        var temp = min_inclusive;
+        min_inclusive = max_inclusive;
+        max_inclusive = temp;
+    }
+    return value < min_inclusive ? min_inclusive : value < max_inclusive ? value : max_inclusive;
+};
+
+/**
+ * converts degrees to radians
```
Comment (pandamicro) [inline cocos/core/utils/misc.ts]: Please complete the Chinese documentation

File: cocos/core/utils/misc.ts [hunk 4]
```diff
@@ -186,6 +187,52 @@ export function cloneable_DEV (obj) {
...
+ * @param {Number} angle
+ * @return {Number}
+ * @method degreesToRadians
+ */
+export function degreesToRadians (angle) {
+    return angle * macro.RAD;
+};
+
+/**
+ * converts radians to degrees
```
Comment (pandamicro) [inline cocos/core/utils/misc.ts]: Same

File: cocos/particle/2d/motion-streak-2d.ts [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+ * You construct it with a fadeTime, minimum segment size, texture path, texture            <br/>
+ * length and color. The fadeTime controls how long it takes each vertex in                 <br/>
+ * the streak to fade out, the minimum segment size it how many pixels the                  <br/>
+ * streak will move before adding a new ribbon segment, and the texture                     <br/>
+ * length is the how many pixels the texture is stretched across. The texture               <br/>
+ * is vertically aligned along the streak segment.
+ * !#zh 运动轨迹，用于游戏对象的运动轨迹上实现拖尾渐隐效果。
+ * @class MotionStreak
+ * @extends Component
+ * @uses BlendFunc
```
Comment (pandamicro) [inline cocos/particle/2d/motion-streak-2d.ts]: Check for all comment format

File: cocos/core/assets/index.ts [hunk 6]
```diff
@@ -55,3 +55,4 @@ export { Mesh, RenderingSubMesh } from './mesh';
 export { Skeleton } from './skeleton';
 export { RenderTexture } from './render-texture';
 import './deprecation';
+export { ParticleAsset } from './particle-asset';
```
Comment (pandamicro) [inline cocos/core/assets/index.ts]: Can we put this file to `physics-2d` module?

**Review by pandamicro — CHANGES_REQUESTED** [review]

File: cocos/particle/index.ts (hunk 1 above)
Comment (caryliu1999) [reply cocos/particle/index.ts]: > Please separate to `core/particle-2d` module

Fixed.

File: cocos/core/utils/misc.ts (hunk 2 above)
Comment (caryliu1999) [reply cocos/core/utils/misc.ts]: > Please use correct document comment format

Fixed.

File: cocos/core/assets/index.ts (hunk 6 above)
Comment (caryliu1999) [reply cocos/core/assets/index.ts]: > Can we put this file to `physics-2d` module?

Moved to particle-2d.

File: cocos/core/utils/misc.ts [hunk 7]
```diff
@@ -186,6 +187,53 @@ export function cloneable_DEV (obj) {
...
+ * @en Clamp a value between from and to.
+ * @zh 限定浮点数的最大最小值。<br/>
+ * 数值大于 max_inclusive 则返回 max_inclusive。<br/>
+ * 数值小于 min_inclusive 则返回 min_inclusive。<br/>
+ * 否则返回自身。
+ * @method clampf
+ * @param {Number} value
+ * @param {Number} min_inclusive
+ * @param {Number} max_inclusive
+ * @return {Number}
```
Comment (pandamicro) [inline cocos/core/utils/misc.ts:199]: Type is not needed

File: cocos/core/utils/misc.ts [hunk 8]
```diff
@@ -186,6 +187,53 @@ export function cloneable_DEV (obj) {
         ((obj.constructor && obj.constructor.prototype.hasOwnProperty('clone')) || obj.hasOwnProperty('clone'));
 }
 
+/**
+ * @en Clamp a value between from and to.
+ * @zh 限定浮点数的最大最小值。<br/>
+ * 数值大于 max_inclusive 则返回 max_inclusive。<br/>
+ * 数值小于 min_inclusive 则返回 min_inclusive。<br/>
+ * 否则返回自身。
+ * @method clampf
```
Comment (pandamicro) [inline cocos/core/utils/misc.ts]: method is not needed

File: cocos/particle-2d/motion-streak-2d.ts [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+ * You construct it with a fadeTime, minimum segment size, texture path, texture            <br/>
+ * length and color. The fadeTime controls how long it takes each vertex in                 <br/>
+ * the streak to fade out, the minimum segment size it how many pixels the                  <br/>
+ * streak will move before adding a new ribbon segment, and the texture                     <br/>
+ * length is the how many pixels the texture is stretched across. The texture               <br/>
+ * is vertically aligned along the streak segment.
+ * @zh 运动轨迹，用于游戏对象的运动轨迹上实现拖尾渐隐效果。
+ * @class MotionStreak
+ * @extends Component
+ * @uses BlendFunc
```
Comment (pandamicro) [inline cocos/particle-2d/motion-streak-2d.ts]: class, extends, uses are all useless in the new doc system

File: cocos/particle-2d/motion-streak-2d.ts [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+@menu('UI/Render/MotionStreak')
+@help('i18n:COMPONENT.help_url.motionStreak')
+export class MotionStreak extends UIRenderable {
+    @serializable
+    private _preview: boolean = false;
+
+    /**
+     * @en Preview the trailing effect in editor mode.
+     * @zh 在编辑器模式下预览拖尾效果。
+     * @property {Boolean} preview
```
Comment (pandamicro) [inline cocos/particle-2d/motion-streak-2d.ts]: name of property is automatically identified

File: cocos/particle-2d/motion-streak-2d.ts [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+        this.reset();
+    }
+
+    @serializable
+    private _minSeg = 1;
+    /**
+     * @en The minimum segment size.
+     * @zh 拖尾之间最小距离。
+     * @property minSeg
+     * @type {Number}
```
Comment (pandamicro) [inline cocos/particle-2d/motion-streak-2d.ts]: type also identified with typescript

Comment (caryliu1999) [reply cocos/particle-2d/motion-streak-2d.ts]: > type also identified with typescript

Fixed.

Comment (pandamicro) [reply cocos/particle-2d/motion-streak-2d.ts]: Didn't see any fix

**Review by pandamicro — APPROVED** [review]

**Pull request merged**